### PR TITLE
Water-2803: Charge Information - Create custom charge

### PR DESCRIPTION
**Pull request by Stephan-Hough**
Feat/water 2803 add custom charge

### CODE REVIEW

File: test/internal/modules/charge-information/controllers/charge-element.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+const routingConfig = {
+  purpose: { pageTitle: 'Select a purpose use', nextStep: 'description' },
+  description: { pageTitle: 'Add element description', nextStep: 'abstraction', back: 'purpose' },
+  abstraction: { pageTitle: 'Set abstraction period', nextStep: 'quantities', back: 'description' },
+  quantities: { pageTitle: 'Add licence quantities', nextStep: 'time', back: 'abstraction' },
+  time: { pageTitle: 'Set time limit?', nextStep: 'source', back: 'quantities' },
+  source: { pageTitle: 'Select source', nextStep: 'season', back: 'time' },
+  season: { pageTitle: 'Select season', nextStep: 'loss', back: 'source' },
+  loss: { pageTitle: 'Select loss category', nextStep: 'loss', back: 'season' }
+};
```
Comment (DGajewska) [inline test/internal/modules/charge-information/controllers/charge-element.js:27]: If you require this from the controller, or put it in it's own file, it will make it easier when making changes to this flow in future

File: test/internal/modules/charge-information/controllers/charge-element.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  },
+  server: {
+    methods: {
+      setDraftChargeInformation: sandbox.stub()
+    }
+  }
+});
+
+// const getReadableDate = str => moment(str).format('D MMMM YYYY');
+// const getISODate = str => moment(str).format('YYYY-MM-DD');
```
Comment (DGajewska) [inline test/internal/modules/charge-information/controllers/charge-element.js:71]: Some comments you forgot to remove

File: test/internal/modules/charge-information/routes/charge-element.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+const validChargeElementSteps = [
+  'purpose',
+  'description',
+  'abstraction',
+  'quantities',
+  'time',
+  'source',
+  'season',
+  'loss'
+];
```
Comment (DGajewska) [inline test/internal/modules/charge-information/routes/charge-element.js:25]: As mentioned in another test, if you take these as the keys from the `routingConfig` object in the controller, this makes future changes easier

**Review by DGajewska — CHANGES_REQUESTED** [review]

File: src/internal/modules/charge-information/lib/charge-elements/data-service.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+'use-strict';
+
+const sessionHelpers = require('shared/lib/session-helpers');
+
+const saveCustomCharge = (request, licenceId, elementId, sessionData) => {
+  const { draftChargeInformation } = request.pre;
+  // add the charge element to the draft charge information
+  if (draftChargeInformation.chargeElements) {
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/lib/charge-elements/data-service.js:8]: Great thanks - The chargeInformationPlugin set the chargeElements = false so I changed it to initialise an empty array so i then removed the unnecessary if as suggested.

File: src/internal/modules/charge-information/controllers/charge-element.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
+'use-strict';
+
+const sessionForms = require('shared/lib/session-forms');
+const formHelpers = require('shared/lib/forms');
+const forms = require('../forms/charge-element/index');
+const { omit } = require('lodash');
+const mappers = require('../lib/charge-elements/mappers');
+const dataservice = require('../lib/charge-elements/data-service');
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/controllers/charge-element.js:8]: done

File: src/internal/modules/charge-information/controllers/charge-element.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ */
+const routingConfig = {
+  purpose: { pageTitle: 'Select a purpose use', nextStep: 'description' },
+  description: { pageTitle: 'Add element description', nextStep: 'abstraction', back: 'purpose' },
+  abstraction: { pageTitle: 'Set abstraction period', nextStep: 'quantities', back: 'description' },
+  quantities: { pageTitle: 'Add licence quantities', nextStep: 'time', back: 'abstraction' },
+  time: { pageTitle: 'Set time limit?', nextStep: 'source', back: 'quantities' },
+  source: { pageTitle: 'Select source', nextStep: 'season', back: 'time' },
+  season: { pageTitle: 'Select season', nextStep: 'loss', back: 'source' },
+  loss: { pageTitle: 'Select loss category', nextStep: 'loss', back: 'season' }
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/controllers/charge-element.js]: Thanks

File: src/internal/modules/charge-information/forms/charge-element/loss-category.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
+'use strict';
+
+const routing = require('../../lib/routing');
+const Joi = require('@hapi/joi');
+const { formFactory, fields } = require('shared/lib/forms/');
+const { has } = require('lodash');
+
+const options = (defaultChargeData, selectedPurpose) => {
+  const defaultLoss = defaultChargeData.find(row => row.purposeUse.id === selectedPurpose.id);
+  const loss = has(defaultLoss, 'loss') ? defaultLoss.loss : 'high';
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/forms/charge-element/loss-category.js:11]: Thanks

File: src/internal/modules/charge-information/forms/charge-element/quantities.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+ * @param {Boolean}  data object containing selected and default options for the form
+  */
+const form = (request, sessionData = {}) => {
+  const { csrfToken } = request.view;
+  const { licenceId, elementId } = request.params;
+  const action = routing.getChargeElementStep(licenceId, elementId, 'quantities');
+
+  const f = formFactory(action, 'POST');
+  f.fields.push(getFormField('authorised', sessionData));
+  f.fields.push(getFormField('billable', sessionData));
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/forms/charge-element/quantities.js:38]: done

File: src/internal/modules/charge-information/forms/charge-element/season.js [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  const action = routing.getChargeElementStep(licenceId, elementId, 'season');
+  const f = formFactory(action, 'POST');
+
+  f.fields.push(fields.radio('season', {
+    errors: {
+      'any.required': {
+        message: 'Select a season'
+      }
+    },
+    choices: options(sessionData.abstractionPeriod || null)
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/forms/charge-element/season.js:37]: removed

File: src/internal/modules/charge-information/forms/charge-element/time-limit.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  f.fields.push(fields.radio('timeLimitedPeriod', {
+    errors: {
+      'any.required': {
+        message: 'Select yes if you want to set a time limit. Select no to continue'
+      }
+    },
+    choices: options(sessionData)
+  }, getSelectedValue(sessionData)));
+  f.fields.push(fields.hidden('chargeStartDate', {}, draftChargeInformation.startDate));
+  f.fields.push(fields.hidden('expiredDate', {}, licence.expiredDate || '9999-01-01'));
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/forms/charge-element/time-limit.js:99]: removed - thanks

File: src/internal/modules/charge-information/lib/charge-elements/mappers.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
+'use-strict';
+
+const abstraction = (formValues) => {
+  const [startMonth, startDay] = (formValues.startDate || '').toString().split(/[- T]/g);
+  const [endMonth, endDay] = (formValues.endDate || '').toString().split(/[- T]/g);
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/lib/charge-elements/mappers.js:5]: I removed the ` || '' ` because the form validation will prevent the start and end date from being empty.

File: src/internal/modules/charge-information/lib/charge-elements/mappers.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+const time = (formValues) => {
+  if (formValues.timeLimitedPeriod === 'yes') {
+    return {
+      timeLimitedPeriod: {
+        startDate: formValues.startDate,
+        endDate: formValues.endDate
+      }
+    };
+  } else return { timeLimitedPeriod: false };
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/charge-information/lib/charge-elements/mappers.js:24]: thanks changed to null - I thought the view needed it to be false but this is not the case

File: src/shared/lib/forms/mappers.js [hunk 13]
```diff
@@ -98,6 +99,40 @@ const dateMapper = {
...
+ * This discussion thread includes further details https://github.com/sideway/joi/issues/1245
+ */
+const dayOfYearMapper = {
+  import: (fieldName, payload) => {
+    const day = payload[fieldName + '-day'];
+    const month = payload[fieldName + '-month'];
+    const date = moment(`2001-${formatDateSegment(month)}-${formatDateSegment(day)}`);
+    // add the year if it is a valid date otherwise add invalid to prevent Javascript to convert this as a date that might pass the Joi.date() validation
+    const value = (date.isValid()) ? `2001-${formatDateSegment(month)}-${formatDateSegment(day)}` : `invalid${formatDateSegment(month)}-${formatDateSegment(day)}`;
+    return value;
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/shared/lib/forms/mappers.js:116]: Thanks

File: src/shared/view/nunjucks/filters/form.js [hunk 14]
```diff
@@ -54,26 +54,16 @@ const mapFormField = (field) => {
...
-      name: 'year'
-    }
-  ];
+const getFormDateItems = (values, items = ['day', 'month', 'year']) => {
+  const formValues = (values || '').toString().split(/[- T]/g);
+  return items.map((item, index) => {
+    return {
+      classes: item === 'year' ? 'govuk-input--width-4' : 'govuk-input--width-2',
+      // items.length -1 because the array starts at 0 not 1
+      value: formValues[(items.length - 1) - index],
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/shared/view/nunjucks/filters/form.js:63]: I see what you are saying but this will cause further problems up the chain because the month will be provided first if the year is not present so the month will be mapped to the year and day will be mapped to the year. For the abstraction period we expect the day-month not the month-day.

File: test/internal/modules/charge-information/controllers/charge-element.js (hunk 1 above)
Comment (Stephan-Hough) [reply test/internal/modules/charge-information/controllers/charge-element.js:27]: done thanks

**Review by DGajewska — APPROVED** [review]
